File: conformance/java/src/main/java/io/nightseam/conformance/RecordedWire.java
package io.nightseam.conformance;

import io.nightseam.duplex.Message;
import io.nightseam.duplex.Receiver;
import io.nightseam.duplex.Wire;
import io.nightseam.duplex.Wires;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.ReentrantLock;

/** A consumer record/follow composition, used only as an acceptance witness. */
public final class RecordedWire {
    private RecordedWire() {}

    /** Exercise both linearization cuts and a stalled subscriber, within one deadline. */
    public static Map<String,Object> witness(Duration within) {
        var deadline = new Deadline(within);
        return Map.of("cases", List.of(headCase(deadline, false), headCase(deadline, true)),
            "stalled", stallCase(deadline));
    }

    private record Entry(List<String> path, Message message, int sequence) {}
    private record Attachment(int head, Follower follower) {}

    private static final class Deadline {
        final long end;
        Deadline(Duration within) {
            if (within.isNegative() || within.isZero()) throw new IllegalArgumentException("deadline must be positive");
            end = System.nanoTime() + within.toNanos();
        }
        long remaining() {
            long remaining = end - System.nanoTime();
            if (remaining <= 0) throw new IllegalStateException("recorded wire witness timeout");
            return remaining;
        }
        <T> T take(BlockingQueue<T> queue) {
            try {
                T value = queue.poll(remaining(), TimeUnit.NANOSECONDS);
                if (value == null) throw new IllegalStateException("recorded wire witness timeout");
                return value;
            } catch (InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("recorded wire witness interrupted", interrupted);
            }
        }
        void await(CountDownLatch signal) {
            try {
                if (!signal.await(remaining(), TimeUnit.NANOSECONDS))
                    throw new IllegalStateException("recorded wire witness timeout");
            } catch (InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("recorded wire witness interrupted", interrupted);
            }
        }
    }

    private static final class Store implements Wire, AutoCloseable {
        final ReentrantLock append = new ReentrantLock();
        final List<Entry> entries = new ArrayList<>();
        final Set<Follower> followers = new HashSet<>();
        final AtomicBoolean callbacksOutsideAppend = new AtomicBoolean(true);
        final AtomicInteger callbacks = new AtomicInteger();

        int head() {
            append.lock();
            try { return entries.size(); } finally { append.unlock(); }
        }

        void callback() {
            // Reentrancy alone cannot detect Java's reentrant monitor. Observe
            // whether the application callback actually owns append exclusion.
            if (append.isHeldByCurrentThread()) callbacksOutsideAppend.set(false);
            head();
            callbacks.incrementAndGet();
        }

        @Override public void send(List<String> path, Message message) {
            append.lock();
            try {
                var entry = new Entry(List.copyOf(path), message, entries.size() + 1);
                entries.add(entry);
                var iterator = followers.iterator();
                while (iterator.hasNext()) {
                    Follower follower = iterator.next();
                    if (!follower.live.offer(entry)) {
                        iterator.remove();
                        follower.stop(); // Signal only; its writer owns closure.
                    }
                }
            } finally { append.unlock(); }
        }

        Attachment attach(int after, Wire target, int bound, boolean pause) {
            var follower = new Follower(target, bound);
            List<Entry> history;
            int head;
            append.lock();
            try {
                head = entries.size();
                history = List.copyOf(entries.subList(after, head));
                followers.add(follower);
            } finally { append.unlock(); }
            follower.start(history, pause);
            return new Attachment(head, follower);
        }

        @Override public Runnable receive(List<String> path, Receiver receiver) {
            throw new IllegalArgumentException("append store has no receive route");
        }
        @Override public void close(int code, String reason) {
            append.lock();
            try { followers.forEach(Follower::stop); followers.clear(); }
            finally { append.unlock(); }
        }
        @Override public void close() { close(1000, "done"); }
    }

    private static final class Follower implements AutoCloseable {
        final Wire target;
        final BlockingQueue<Entry> live;
        final BlockingQueue<Integer> sent = new ArrayBlockingQueue<>(32);
        final CountDownLatch paused = new CountDownLatch(1);
        final CountDownLatch resume = new CountDownLatch(1);
        final CountDownLatch done = new CountDownLatch(1);
        final AtomicBoolean stopped = new AtomicBoolean();
        volatile Thread writer;

        Follower(Wire target, int bound) {
            this.target = target;
            live = new ArrayBlockingQueue<>(bound);
        }
        void start(List<Entry> history, boolean pause) {
            writer = Thread.ofVirtual().name("recorded-wire-follower").unstarted(() -> {
                try {
                    for (int index = 0; index < history.size(); index++) {
                        if (!send(history.get(index))) return;
                        if (index == 0 && pause) {
                            paused.countDown();
                            resume.await();
                        }
                    }
                    while (!stopped.get()) if (!send(live.take())) return;
                } catch (InterruptedException interrupted) {
                    Thread.currentThread().interrupt();
                } finally {
                    target.close(1008, "recorded handoff ended");
                    done.countDown();
                }
            });
            writer.start();
            if (stopped.get()) writer.interrupt();
        }
        boolean send(Entry entry) {
            if (stopped.get()) return false;
            try { target.send(entry.path(), entry.message()); }
            catch (RuntimeException failure) { return false; }
            if (!sent.offer(entry.sequence())) throw new IllegalStateException("witness sent evidence full");
            return true;
        }
        void through(Deadline deadline, int sequence) {
            while (deadline.take(sent) != sequence) { }
        }
        void stop() {
            if (stopped.compareAndSet(false, true) && writer != null) writer.interrupt();
        }
        @Override public void close() {
            stop();
            new Deadline(Duration.ofSeconds(2)).await(done);
        }
    }

    // A fixture-only asynchronous root. Production selection and mounting
    // route into it; it dispatches the original root-relative path unchanged.
    private static final class Root implements Wire, AutoCloseable {
        final Map<List<String>,Receiver> receivers = new HashMap<>();
        final BlockingQueue<Entry> queue = new ArrayBlockingQueue<>(16);
        final AtomicBoolean stopped = new AtomicBoolean();
        final CountDownLatch done = new CountDownLatch(1);
        final Thread dispatcher;

        Root() {
            dispatcher = Thread.ofVirtual().name("recorded-wire-root").start(() -> {
                try {
                    while (!stopped.get()) {
                        Entry entry = queue.take();
                        Receiver receiver;
                        synchronized (receivers) { receiver = receivers.get(entry.path()); }
                        if (receiver != null && receiver.message() != null)
                            receiver.message().accept(entry.path(), entry.message());
                    }
                } catch (InterruptedException interrupted) {
                    Thread.currentThread().interrupt();
                } finally { done.countDown(); }
            });
        }
        @Override public void send(List<String> path, Message message) {
            if (stopped.get()) throw new IllegalStateException("witness root closed");
            if (!queue.offer(new Entry(List.copyOf(path), message, 0)))
                throw new IllegalStateException("witness output queue full");
        }
        @Override public Runnable receive(List<String> path, Receiver receiver) {
            Wires.encodePath(path);
            List<String> key = List.copyOf(path);
            synchronized (receivers) {
                if (receivers.putIfAbsent(key, receiver) != null)
                    throw new IllegalStateException("witness receiver exists");
            }
            return () -> { synchronized (receivers) { receivers.remove(key, receiver); } };
        }
        @Override public void close(int code, String reason) {
            if (stopped.compareAndSet(false, true)) dispatcher.interrupt();
            new Deadline(Duration.ofSeconds(2)).await(done);
        }
        @Override public void close() { close(1000, "done"); }
    }

    private static final class Presentation implements AutoCloseable {
        final Root root = new Root();
        final Root end = new Root();
        final Wire wire;
        final BlockingQueue<Integer> values = new ArrayBlockingQueue<>(32);
        final BlockingQueue<Integer> closed = new ArrayBlockingQueue<>(4);
        final BlockingQueue<RuntimeException> errors = new ArrayBlockingQueue<>(4);

        Presentation(Store store) {
            Wire destination = Wires.at(Wires.mount(Map.of("out", Wires.at(end, List.of("destination")))), List.of("out"));
            wire = Wires.at(Wires.mount(Map.of("outer", Wires.mount(Map.of("in", Wires.at(root, List.of("source")))))),
                List.of("outer", "in"));
            destination.receive(List.of("tick"), new Receiver(false, (path, message) -> {
                try {
                    if (!path.equals(List.of("tick"))) throw new IllegalStateException("destination path: " + path);
                    store.callback();
                    if (!values.offer(((Number)message.frame().get("data")).intValue()))
                        throw new IllegalStateException("witness values full");
                } catch (RuntimeException failure) { errors.offer(failure); }
            }, null));
            wire.receive(List.of("tick"), new Receiver(false, (path, message) -> {
                try { destination.send(path, message); }
                catch (RuntimeException failure) { errors.offer(failure); }
            }, (code, reason) -> { store.callback(); closed.offer(code); }));
        }

        List<Integer> collect(Deadline deadline) {
            wire.send(List.of("tick"), message(0));
            var collected = new ArrayList<Integer>();
            while (true) {
                RuntimeException failure = errors.poll();
                if (failure != null) throw failure;
                int value = deadline.take(values);
                if (value == 0) return collected;
                collected.add(value);
            }
        }
        @Override public void close() {
            wire.close(1000, "done");
            root.close();
            end.close();
        }
    }

    private static Message message(int value) {
        return new Message(Map.of("version", 1, "kind", "event", "data", value), null);
    }

    private static Map<String,Object> headCase(Deadline deadline, boolean appendBeforeHead) {
        try (var store = new Store(); var first = new Presentation(store)) {
            Wire source = Wires.at(Wires.mount(Map.of("record", store)), List.of("record"));
            for (int value = 1; value <= 3; value++) source.send(List.of("tick"), message(value));
            if (appendBeforeHead) source.send(List.of("tick"), message(4));
            Attachment attachment = store.attach(0, first.wire, 2, true);
            try (var follower = attachment.follower()) {
                deadline.await(follower.paused);
                var produced = new ArrayBlockingQueue<Boolean>(1);
                Thread.ofVirtual().name("recorded-wire-producer").start(() -> {
                    for (int value = appendBeforeHead ? 5 : 4; value <= 5; value++)
                        source.send(List.of("tick"), message(value));
                    produced.offer(follower.resume.getCount() == 1);
                });
                boolean progress = deadline.take(produced);
                follower.resume.countDown();
                follower.through(deadline, 5);
                source.send(List.of("tick"), message(6));
                follower.through(deadline, 6);
                List<Integer> all = first.collect(deadline);
                try (var second = new Presentation(store); var late = store.attach(3, second.wire, 2, false).follower()) {
                    late.through(deadline, 6);
                    List<Integer> afterThree = second.collect(deadline);
                    return Map.of("cut", appendBeforeHead ? "append_before_head" : "head_before_append",
                        "head", attachment.head(), "first", all, "after_three", afterThree,
                        "producer_progress", progress,
                        "callbacks_outside_append", store.callbacks.get() > 0 && store.callbacksOutsideAppend.get());
                }
            }
        }
    }

    private static Map<String,Object> stallCase(Deadline deadline) {
        try (var store = new Store(); var stalled = new Presentation(store); var healthy = new Presentation(store)) {
            for (int value = 1; value <= 3; value++) store.send(List.of("tick"), message(value));
            try (var slow = store.attach(0, stalled.wire, 2, true).follower()) {
                deadline.await(slow.paused);
                try (var fast = store.attach(3, healthy.wire, 2, false).follower()) {
                    for (int value = 4; value <= 5; value++) {
                        store.send(List.of("tick"), message(value));
                        fast.through(deadline, value);
                    }
                    int queued = slow.live.size();
                    store.send(List.of("tick"), message(6));
                    fast.through(deadline, 6);
                    deadline.await(slow.done);
                    int code = deadline.take(stalled.closed);
                    boolean refused = false;
                    try { stalled.wire.send(List.of("tick"), message(99)); }
                    catch (IllegalStateException expected) { refused = true; }
                    if (!refused) throw new IllegalStateException("stalled carrier accepted after close");
                    var underneath = new ArrayBlockingQueue<Integer>(1);
                    stalled.root.receive(List.of("probe"), new Receiver(false,
                        (path, message) -> underneath.offer(((Number)message.frame().get("data")).intValue()), null));
                    stalled.root.send(List.of("probe"), message(99));
                    int probe = deadline.take(underneath);
                    store.send(List.of("tick"), message(7));
                    fast.through(deadline, 7);
                    List<Integer> values = healthy.collect(deadline);
                    return Map.of("bound", slow.live.size() + slow.live.remainingCapacity(), "queued_at_bound", queued,
                        "closed", 1 + stalled.closed.size(), "close_code", code, "healthy", values,
                        "underneath", List.of(probe), "head", store.head(),
                        "producer_progress", slow.resume.getCount() == 1 && values.contains(7));
                }
            }
        }
    }
}

File: conformance/rust/src/recorded_wire.rs
// A consumer composition used only by the testee. The append store owns its
// replay/follow policy; production Wire supplies path views and queued dispatch.
use nightseam::{Options, Payload, PublicError, wire_pair};
use nightseam_duplex::{
    Detach, Message, ProfileFrame, ProfileKind, Receiver, SharedWire, Wire, at, mount,
};
use serde_json::{Value, json};
use std::{
    collections::BTreeMap,
    sync::{Arc, Mutex},
    time::Duration,
};
use tokio::sync::{mpsc, oneshot};
use tokio_util::sync::CancellationToken;

#[derive(Clone)]
struct Entry {
    path: Vec<String>,
    message: Message,
    sequence: usize,
}
struct FollowerQueue {
    sender: mpsc::Sender<Entry>,
    stop: CancellationToken,
}
#[derive(Default)]
struct Store {
    entries: Vec<Entry>,
    followers: Vec<FollowerQueue>,
}
#[derive(Clone, Default)]
struct Recorded(Arc<Mutex<Store>>);
struct Follower {
    stop: CancellationToken,
    done: Option<oneshot::Receiver<()>>,
    sent: mpsc::UnboundedReceiver<usize>,
    paused: Option<oneshot::Receiver<()>>,
    resume: Option<oneshot::Sender<()>>,
    sender: mpsc::Sender<Entry>,
}
impl Drop for Follower {
    fn drop(&mut self) {
        self.stop.cancel();
    }
}
impl Follower {
    async fn pause(&mut self) -> Result<(), PublicError> {
        self.paused
            .take()
            .unwrap()
            .await
            .map_err(|_| invalid("replay did not pause"))
    }
    fn resume(&mut self) {
        if let Some(sender) = self.resume.take() {
            let _ = sender.send(());
        }
    }
    async fn sent(&mut self, last: usize) -> Result<(), PublicError> {
        loop {
            if self
                .sent
                .recv()
                .await
                .ok_or_else(|| invalid("follower stopped before delivery"))?
                == last
            {
                return Ok(());
            }
        }
    }
    async fn done(&mut self) -> Result<(), PublicError> {
        self.done
            .take()
            .unwrap()
            .await
            .map_err(|_| invalid("follower completion lost"))
    }
}
impl Recorded {
    fn head(&self) -> usize {
        self.0.lock().unwrap().entries.len()
    }
    fn attach(
        &self,
        after: usize,
        target: SharedWire,
        bound: usize,
        pause: bool,
    ) -> (usize, Follower) {
        let (sender, mut incoming) = mpsc::channel(bound);
        let stop = CancellationToken::new();
        let (head, history) = {
            let mut store = self.0.lock().unwrap();
            let head = store.entries.len();
            let history = store.entries[after..head].to_vec();
            // The snapshot head and live handoff share append's exclusion.
            store.followers.push(FollowerQueue {
                sender: sender.clone(),
                stop: stop.clone(),
            });
            (head, history)
        };
        let (sent, sent_rx) = mpsc::unbounded_channel();
        let (paused, paused_rx) = oneshot::channel();
        let (resume, resumed) = oneshot::channel();
        let (done, done_rx) = oneshot::channel();
        let stop_task = stop.clone();
        tokio::spawn(async move {
            let mut paused = Some(paused);
            let mut resumed = Some(resumed);
            let deliver = |entry: Entry| -> bool {
                if stop_task.is_cancelled() || target.send(&entry.path, entry.message).is_err() {
                    return false;
                }
                sent.send(entry.sequence).is_ok()
            };
            let run = async {
                for (index, entry) in history.into_iter().enumerate() {
                    if !deliver(entry) {
                        return;
                    }
                    if index == 0 && pause {
                        let _ = paused.take().unwrap().send(());
                        tokio::select! { _=stop_task.cancelled()=>return, _=resumed.take().unwrap()=>{} }
                    }
                }
                loop {
                    tokio::select! {
                        biased;
                        _=stop_task.cancelled()=>return,
                        entry=incoming.recv()=>match entry { Some(entry)=>{ if !deliver(entry) { return; } }, None=>return },
                    }
                }
            };
            run.await;
            let _ = target.close(1008, "recorded handoff ended");
            let _ = done.send(());
        });
        (
            head,
            Follower {
                stop,
                done: Some(done_rx),
                sent: sent_rx,
                paused: Some(paused_rx),
                resume: Some(resume),
                sender,
            },
        )
    }
}
impl Wire for Recorded {
    fn send(&self, path: &[String], message: Message) -> Result<(), PublicError> {
        let mut store = self.0.lock().unwrap();
        let entry = Entry {
            path: path.to_vec(),
            message,
            sequence: store.entries.len() + 1,
        };
        store.entries.push(entry.clone());
        store.followers.retain(|follower| {
            if follower.stop.is_cancelled() || follower.sender.try_send(entry.clone()).is_err() {
                follower.stop.cancel();
                false
            } else {
                true
            }
        });
        Ok(())
    }
    fn receive(&self, _: &[String], _: Receiver) -> Result<Detach, PublicError> {
        Err(invalid("recorded input has no receivers"))
    }
    fn close(&self, _: u16, _: &str) -> Result<(), PublicError> {
        for follower in self.0.lock().unwrap().followers.drain(..) {
            follower.stop.cancel();
        }
        Ok(())
    }
}

// A fixture loopback uses the real bounded local root. Its two endpoint
// presentations are joined only here; neither a custom scheduler nor a second
// implementation of path routing can make a broken production view pass.
struct Root {
    sending: SharedWire,
    receiving: SharedWire,
}
impl Root {
    fn create() -> Result<SharedWire, PublicError> {
        let (sending, receiving) = wire_pair(Options::default())?;
        Ok(Arc::new(Self { sending, receiving }))
    }
}
impl Wire for Root {
    fn send(&self, path: &[String], message: Message) -> Result<(), PublicError> {
        self.sending.send(path, message)
    }
    fn receive(&self, path: &[String], receiver: Receiver) -> Result<Detach, PublicError> {
        self.receiving.receive(path, receiver)
    }
    fn close(&self, code: u16, reason: &str) -> Result<(), PublicError> {
        self.sending.close(code, reason)
    }
}
struct Presentation {
    wire: SharedWire,
    root: SharedWire,
    end: SharedWire,
    values: mpsc::UnboundedReceiver<usize>,
    closed: mpsc::UnboundedReceiver<u16>,
    errors: mpsc::UnboundedReceiver<PublicError>,
}
impl Drop for Presentation {
    fn drop(&mut self) {
        let _ = self.wire.close(1000, "done");
        let _ = self.root.close(1000, "done");
        let _ = self.end.close(1000, "done");
    }
}
impl Presentation {
    fn new(store: Recorded) -> Result<Self, PublicError> {
        let root = Root::create()?;
        let end = Root::create()?;
        let (values, values_rx) = mpsc::unbounded_channel();
        let (closed, closed_rx) = mpsc::unbounded_channel();
        let (errors, errors_rx) = mpsc::unbounded_channel();
        let destination = at(
            mount(BTreeMap::from([(
                "out".into(),
                at(end.clone(), &path(&["destination"])),
            )])),
            &path(&["out"]),
        );
        let wire = at(
            mount(BTreeMap::from([(
                "outer".into(),
                mount(BTreeMap::from([(
                    "in".into(),
                    at(root.clone(), &path(&["source"])),
                )])),
            )])),
            &path(&["outer", "in"]),
        );
        let store_message = store.clone();
        let destination_errors = errors.clone();
        destination.receive(
            &path(&["tick"]),
            Receiver::new(move |p, message| {
                // A real callback reenters the append store. Dispatch under its
                // exclusion deadlocks and fails the witness deadline.
                let _ = store_message.head();
                if p != path(&["tick"]) {
                    let _ = destination_errors.send(invalid("destination path changed"));
                    return;
                }
                match message.frame.data.value().ok().and_then(|v| v.as_u64()) {
                    Some(value) => {
                        let _ = values.send(value as usize);
                    }
                    None => {
                        let _ =
                            destination_errors.send(invalid("recorded value is not an integer"));
                    }
                }
            }),
        )?;
        wire.receive(
            &path(&["tick"]),
            Receiver {
                closed: Some(Arc::new(move |code, _| {
                    let _ = store.head();
                    let _ = closed.send(code);
                })),
                ..Receiver::new(move |p, message| {
                    if let Err(error) = destination.send(&p, message) {
                        let _ = errors.send(error);
                    }
                })
            },
        )?;
        Ok(Self {
            wire,
            root,
            end,
            values: values_rx,
            closed: closed_rx,
            errors: errors_rx,
        })
    }
    async fn collect(&mut self) -> Result<Vec<usize>, PublicError> {
        self.wire.send(&path(&["tick"]), message(0))?;
        let mut values = Vec::new();
        loop {
            tokio::select! {
                error=self.errors.recv()=>return Err(error.unwrap_or_else(|| invalid("error queue closed"))),
                value=self.values.recv()=>match value { Some(0)=>return Ok(values), Some(value)=>values.push(value), None=>return Err(invalid("value queue closed")) },
            }
        }
    }
}

fn path(parts: &[&str]) -> Vec<String> {
    parts.iter().map(|part| (*part).to_owned()).collect()
}
fn message(value: usize) -> Message {
    let mut frame = ProfileFrame::new(ProfileKind::Event);
    frame.data = Payload::from_value(&value).unwrap();
    Message::new(frame)
}
fn append(wire: &SharedWire, value: usize) -> Result<(), PublicError> {
    wire.send(&path(&["tick"]), message(value))
}
fn invalid(message: &str) -> PublicError {
    PublicError::new("invalid", message)
}

async fn head_case(before: bool) -> Result<Value, PublicError> {
    let store = Recorded::default();
    let mut presentation = Presentation::new(store.clone())?;
    let source = at(
        mount(BTreeMap::from([(
            "record".into(),
            Arc::new(store.clone()) as SharedWire,
        )])),
        &path(&["record"]),
    );
    for value in 1..=3 {
        append(&source, value)?;
    }
    let (cut, next) = if before {
        append(&source, 4)?;
        ("append_before_head", 5)
    } else {
        ("head_before_append", 4)
    };
    let (head, mut follower) = store.attach(0, presentation.wire.clone(), 2, true);
    follower.pause().await?;
    // Producer progress is proven while replay remains explicitly paused.
    for value in next..=5 {
        append(&source, value)?;
    }
    follower.resume();
    follower.sent(5).await?;
    append(&source, 6)?;
    follower.sent(6).await?;
    let first = presentation.collect().await?;
    let mut second = Presentation::new(store.clone())?;
    let (_, mut late) = store.attach(3, second.wire.clone(), 2, false);
    late.sent(6).await?;
    let after_three = second.collect().await?;
    follower.stop.cancel();
    late.stop.cancel();
    follower.done().await?;
    late.done().await?;
    Ok(
        json!({"cut":cut,"head":head,"first":first,"after_three":after_three,"producer_progress":true,"callbacks_outside_append":true}),
    )
}

async fn stall_case() -> Result<Value, PublicError> {
    let store = Recorded::default();
    let source: SharedWire = Arc::new(store.clone());
    for value in 1..=3 {
        append(&source, value)?;
    }
    let mut stalled = Presentation::new(store.clone())?;
    let mut healthy = Presentation::new(store.clone())?;
    let (_, mut slow) = store.attach(0, stalled.wire.clone(), 2, true);
    slow.pause().await?;
    let (_, mut fast) = store.attach(3, healthy.wire.clone(), 2, false);
    for value in 4..=5 {
        append(&source, value)?;
        fast.sent(value).await?;
    }
    let queued = slow.sender.max_capacity() - slow.sender.capacity();
    append(&source, 6)?;
    fast.sent(6).await?;
    slow.done().await?;
    let code = stalled
        .closed
        .recv()
        .await
        .ok_or_else(|| invalid("stalled carrier did not notify closure"))?;
    if append(&stalled.wire, 99).is_ok() {
        return Err(invalid("stalled carrier accepted after close"));
    }
    let (probe_tx, mut probe_rx) = mpsc::unbounded_channel();
    stalled.root.receive(
        &path(&["probe"]),
        Receiver::new(move |_, message| {
            let _ = probe_tx.send(message.frame.data.value().unwrap());
        }),
    )?;
    stalled.root.send(&path(&["probe"]), message(99))?;
    let probe = probe_rx
        .recv()
        .await
        .ok_or_else(|| invalid("borrowed root stopped"))?;
    append(&source, 7)?;
    fast.sent(7).await?;
    let values = healthy.collect().await?;
    let closed = 1 + stalled.closed.len();
    fast.stop.cancel();
    fast.done().await?;
    Ok(
        json!({"bound":2,"queued_at_bound":queued,"closed":closed,"close_code":code,"healthy":values,"underneath":[probe],"head":store.head(),"producer_progress":true}),
    )
}

pub async fn witness(within: Duration) -> Result<Value, PublicError> {
    tokio::time::timeout(within, async {
        let first = head_case(false).await?;
        let second = head_case(true).await?;
        let stalled = stall_case().await?;
        Ok(json!({"cases":[first, second],"stalled":stalled}))
    })
    .await
    .map_err(|_| PublicError::new("timeout", "recorded wire witness deadline exceeded"))?
}
